### PR TITLE
Player steer vehicle event

### PR DESCRIPTION
**Pull request by rgnter**
Event called when packet [`PacketPlayInSteerVehicle`](https://wiki.vg/Protocol#Steer_Vehicle) is sent by player.

### CODE REVIEW

**Comment by electronicboy** [comment]
Wrong package, events generally fire before mutation, i also don't see the use of this as it's going to miss a lot of context, etc

**Comment by rgnter** [comment]
Nothing gets mutated, and what do you mean by wrong package?

**Comment by Owen1212055** [comment]
The ``io.papermc.paper`` package should be used instead.

**Comment by lynxplay** [comment]
Your event should live in the io.papernc package. Additionally even without mutation (which idk why this event isn't allowing mutation in the first place) events should still be called before the internal logic happens.

File: patches/api/0352-steer-vehicle-event.patch [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
++package com.destroystokyo.paper.event.player;
++
++import org.bukkit.event.Event;
++import org.bukkit.event.HandlerList;
++import org.jetbrains.annotations.NotNull;
++
++/**
++ * Called when player is steering any vehicle.
++ */
++public class PlayerSteerVehicleEvent extends Event {
```
Comment (Owen1212055) [inline patches/api/0352-steer-vehicle-event.patch:22]: This should extend PlayerEvent, we have no way to actually get the player who steered this.

**Comment by electronicboy** [comment]
> Nothing gets mutated

So, what you're telling me, is that nothing on the server happens at all from this packet?

**Comment by rgnter** [comment]
Alright sorry about that, can fix that. Anyways, if you do not approve of this API extension i am not going to bother. I believe the event should be rather a monitor of action.

File: patches/server/0842-steer-vehicle-event.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ import com.mojang.brigadier.ParseResults;
+@@ -494,6 +495,11 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
+     public void handlePlayerInput(ServerboundPlayerInputPacket packet) {
+         PacketUtils.ensureRunningOnSameThread(packet, this, this.player.getLevel());
+         this.player.setPlayerInput(packet.getXxa(), packet.getZza(), packet.isJumping(), packet.isShiftKeyDown());
++        // paper start - steer vehicle event
++        this.cserver.getPluginManager().callEvent(new PlayerSteerVehicleEvent(
++            packet.getXxa(), packet.getZza(), packet.isJumping(), packet.isShiftKeyDown()
++        ));
++        // paper end
```
Comment (Owen1212055) [inline patches/server/0842-steer-vehicle-event.patch:26]: Use new PlayerSteerVehicleEvent(...).callEvent();
Use a fully qualified import as well.

**Comment by rgnter** [comment]
The values get used by the server, thats right. But since the event is immutable I dont think its really necessary to call it before server logic(which does not mutate these data)

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
Here is my advice here on some minor nit-picks.

**Comment by rgnter** [comment]
Alright, totally forgot about player in this event. I will modify the patch.

**Comment by electronicboy** [comment]
The values get used by the server, which is thus mutating the state of the server

I don't see the need for random events slapped around the server for knowing that a packet was sent by the client, that's more protocol lib territory; I fail to see a useful application of this which wouldn't be better suited for proper API over different mechanisms exposed in the game

**Comment by rgnter** [comment]
Alright, i won't bother modifying the patch if you do not see use of this event.

**Comment by MiniDigger** [comment]
useful application of this would be riding entities cat, I definitely do see a value here, I had to deal with this stuff myself in the past.
I would name the event in a way that its clear that this is the input from the client only.

**Comment by yannicklamprecht** [comment]
> useful application of this would be riding entities cat, I definitely do see a value here, I had to deal with this stuff myself in the past.
> I would name the event in a way that its clear that this is the input from the client only.

PlayerMovementInputEvent?

**Pull request closed without merging**